カレンダーのプログラム(JavaScript)の課題提出

yokomaru opened the pull request:
https://bootcamp.fjord.jp/products/21124

yokomaru committed:
add: minimist to package.json

- 年月指定オプションを実装するため、minimistを追加

yokomaru committed:
add: cal.js

- カレンダー処理を実装
    - オプション指定
    - 今日日付は色を変える
    - 指定した年月でカレンダー表示
    - 指定がない場合実行した日付の年月カレンダーを表示

yokomaru committed:
delete: todayYear and todayMonth

- 上記変数は一度しか使っていないため、変数を使用しないように修正

yokomaru committed:
fix: display calender

- 変数宣言を使用箇所と近い場所にする
- 日付文字列の比較に.toDateStringを使用する
- 不要な変数宣言を削除する

yokomaru committed:
add: error message and return

- エラーメッセージを表示する処理を追加
   - '-y' のみ指定して値がない場合
   - '-m' のみ指定して値がない場合
   - '-y' の指定があり、数字かつ1970~2100の範囲内でない場合
   - '-m' のみ指定して値がない場合
- エラー時にreturn で処理を抜けるため、処理を関数に区切る
    - init() エラーチェックと引数の処理を行う
    - buildCalender() カレンダーの組み立てから表示までを行う

yokomaru committed:
add: minimist in package.json

- package.jsonにminimistを追加

yokomaru committed:
add: chalk in package.json

- package.jsonにchalkを追加

yokomaru committed:
fix: file authority

- `./cal.js`で実行できるようにファイル権限を変更

cafedomancer commented on 02.calendar/package.json:
```diff
@@ -10,5 +10,9 @@
+  "type": "module",
+  "dependencies": {
+    "minimist": "^1.2.8",
+    "chalk": "4.1.2"
```
バージョンが古いです。

yokomaru replied:
` npm view chalk version`で最新バージョンを確認し、`5.3.0`に修正いたしました。

[94c4def](https://github.com/yokomaru/js-practices/pull/2/commits/94c4defc756cefd1396b597fb6c21488a70c5f4d)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+import minimist from "minimist";
+import chalk from "chalk";
+
+function init() {
```
init とは初期化処理のことを指しますが、この関数の中で行われているのは初期化処理ではないです。

yokomaru replied:
コメントいただいた通り、`init` では不適切でした。(カレンダーを表示する前の初期設定みたいな意味合いで使ってしまっていました)

`init`  の中身の処理を見直したときに責務が3つ発生してしまっていたため、
`init` は削除し各々の処理に分割いたしました。

- 値のセット
   - `setMonth`
   - `setYear`

- 値のバリデーション
   - `validateMonth`
   - `validateYear`

- カレンダーの表示の実行

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+  buildCalender(year, month, today);
+}
+
+function buildCalender(year, month, today) {
```
build と付いている関数は普通は何らかの値を生成してそれを返すような関数ですが、そうなっていないです。

yokomaru replied:
実際に行っている処理はカレンダーの表示のため、`printCalender`に修正しました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+  let month = today.getMonth() + 1;
+
+  if (typeof argv["y"] === "boolean")
+    return console.error("cal: option requires an argument -- y");
```
`console.error` は戻り値のない関数なので、その評価結果を return しても意味がないです。

yokomaru replied:
上記承知いたしました。

自分が実装したかったことは実行終了のみで、関数の呼び出し元に値を返したいわけではなかったので、
調べ直して`return`ではなく`process.exit(1)`を使用するように以下で修正いたしました。

[670849b](https://github.com/yokomaru/js-practices/pull/2/commits/670849bd019c1f611cd6b26876decc6c0a445c55)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+  let year = today.getFullYear();
+  let month = today.getMonth() + 1;
+
+  if (typeof argv["y"] === "boolean")
```
オブジェクトのプロパティにアクセスするときはブラケット記法ではなくドット記法を使ってください。

yokomaru replied:
ドット記法を使うように修正いたしました！今後気をつけます🙇‍♀️

[5fa6429](https://github.com/yokomaru/js-practices/pull/2/commits/5fa6429c55312c93343d7cd6b4e9a02259a13d3d)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+  let year = today.getFullYear();
+  let month = today.getMonth() + 1;
+
+  if (typeof argv["y"] === "boolean")
+    return console.error("cal: option requires an argument -- y");
```
if 文は必ずブロックを伴って書いてください。https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Statements/if...else#%E8%A7%A3%E8%AA%AC

yokomaru replied:
修正いたしました！今後気をつけます🙇‍♀️
[24534c4](https://github.com/yokomaru/js-practices/pull/2/commits/24534c4b066bcd84d74fa80167a4c86f32838175)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+  let year = today.getFullYear();
+  let month = today.getMonth() + 1;
+
+  if (typeof argv["y"] === "boolean")
```
質問ですが、`typeof argv["y"] === "boolean"` は何をチェックしているのでしょう？

yokomaru replied:
オプションを指定した際に引数があるかどうかをチェックしています。
`cal -m`を実行すると 表示されるエラーメッセージ(`cal: option requires an argument -- m`)を再現するためです。

実装した処理に`-m`のみを渡したところ、argvの中にtrueが入っていたため、型がbooleanであればエラーメッセージを表示するように実装を行いました。

```
➜  02.calendar git:(my-calender) ✗ ./cal.js -m
{ _: [], m: true }
```

※ 実際のcalコマンドは-yを指定すると12ヶ月分表示されるますが、その挙動を再現するのはプラクティスでは不要なため-mと同様の処理を行っております。

cafedomancer replied:
なるほど、分かりました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+  if (typeof argv["m"] === "boolean")
+    return console.error("cal: option requires an argument -- m");
+
+  if (typeof argv["y"] !== "undefined") {
+    if (
+      typeof argv["y"] === "number" &&
+      1970 <= argv["y"] &&
+      argv["y"] <= 2100
+    ) {
+      year = argv["y"];
+    } else {
+      return console.error(`cal: year \`${argv["y"]}' not in range 1970..2100`);
+    }
+  }
```
条件武器がネストしていて読みづらいです。異常値はガード節で処理してください。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+  const calenderTitle = `${month}月 ${year}`;
+  const calenderTitleLength = calenderTitle.length;
+
+  console.log(
+    `${calenderTitle
+      .padStart(
+        calenderTitleLength + Math.floor((20 - calenderTitleLength) / 2),
+        " ",
+      )
+      .padEnd(20, " ")}`,
+  );
```
長すぎて読みづらく、何が行われているのかも分からないです。

yokomaru replied:
カレンダーのタイトルを1桁月でも2桁月でも同じ位置から表示されるようにセンタリングしたかったため、以下の処理を実装しておりました。

1. calコマンドの全て2桁日付の行(`24 25 26 27 28 29 30`など)の幅が20だったため最大値を20とする
2. 最大値からタイトル(XX月 XXXX年)の長さを引いて余白の合計を求める
3. 余白の合計を左右に均等に割り当てるために2で割り、タイトルの左側に入れるスペースの数を計算する
4. 上記の割り算の結果をMath.floor()で整数値にする
5. 文字列の右側に、全体の長さが20文字になるまでスペースを挿入する

ただ、こちらもう一度改めて確認した際に**タイトルの左側に入れるスペースの数(3.で求めている数字)は1桁月でも2桁月でも6になる**ため、都度複雑な計算をして余白を求める必要がない事に気づきました💦

上記を踏まえてテンプレート文字列でスペースを6つ追加してからタイトルを表示するように修正いたしました。
[59de86d](https://github.com/yokomaru/js-practices/pull/2/commits/59de86dfe43c1f660dc92faeffcd7025eb1c7323)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+
+  console.log("日 月 火 水 木 金 土");
+
+  const firstMonthDate = new Date(year, month - 1, 1);
```
first が month にかかっているのか date にかかっているのか分かりづらいです。

yokomaru replied:
`first`は`date`の方にかけたいので`monthFirstDate`に修正いたしました。
※ 上記変数は実装を見直した際に削除したのですが、命名する際に意識します。
[03c0955](https://github.com/yokomaru/js-practices/pull/2/commits/03c0955d5fdd448406733eb9b3d2df7ab329005c)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+
+  const firstMonthDate = new Date(year, month - 1, 1);
+
+  for (let i = 0; i < new Date(firstMonthDate).getDay(); i++) {
```
質問ですが、何のために `new Date(...)` を実行しているのでしょう？

yokomaru replied:
こちらは`new Date(year, month - 1, 1)`を`firstMonthDate`変数に入れた際の見直し不足でした。
`firstMonthDate`はDateオブジェクトのため、そのまま`.getDay()`が使えるのでnew Date(...) は不要でした。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+
+  for (
+    let date = firstMonthDate;
+    date < new Date(year, month, 1, 0);
```
`new Date(year, month, 1, 0)` が何を表しているのか分からないです。

yokomaru replied:
こちらは翌月の1日を表現しておりました。

UTC時間(+9時間する)をきちんと理解できておらず、月の最終日ではなく翌月1日を比較対象にして
比較演算子`<`を使用して最終日までのループを実装しておりました。
また、Dateオブジェクトを生成する際に、hour の既定値が 0なので指定は不要でした。

以下で最終日を取得するように修正いたしました。
[e2386c2](https://github.com/yokomaru/js-practices/pull/2/commits/e2386c29e388b29c186f60b5737b36fa759a3f5f)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+  for (
+    let date = firstMonthDate;
+    date < new Date(year, month, 1, 0);
+    date.setDate(date.getDate() + 1)
```
`setDate` はオブジェクトそのものの値を書き換えるので、`date` に格納されているオブジェクト、つまり `firstMonthDate` に格納されているオブジェクトの日付も変わってしまいますよ。

yokomaru replied:
コメントありがとうございます。
上記、以下のように理解いたしました。

・変数に変数を代入すると代入された変数の値の参照が格納される
・`setDate`は破壊的な変更を行うため、代入された変数の値自体に変更を加えてしまう

処理を見直した際、元々はDate型オブジェクトを使用してループしようとしておりましたが、
ループして表示する際に必要なのは日にちの数字のため数値でループするように修正いたしました。

[8365217](https://github.com/yokomaru/js-practices/pull/2/commits/8365217ba9c6a696ddfbc8423b5c5ff566a88093)
※ 見返したところ別のコミットも混ざってしまっておりました。。気をつけます🙇‍♀️

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+  ) {
+    const displayDate =
+      today.toDateString() === date.toDateString()
+        ? chalk.bgWhite(`${date.getDate()}`.padStart(2, " "))
```
`` `${date.getDate()}` `` で文字列中に変数が埋め込まれていないのにテンプレート文字列が使われています。

yokomaru replied:
数字を文字列に変換するためにテンプレート文字列を使ってしまっておりました。
`toString()`関数を使い数値から文字列に変更するように修正しました。
重複していたため、変数に入れております。

[7645bc5](https://github.com/yokomaru/js-practices/pull/2/commits/7645bc51b1da9c6ec3e0a322984762bcfc7efaca)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+        ? chalk.bgWhite(`${date.getDate()}`.padStart(2, " "))
+        : `${date.getDate()}`.padStart(2, " ");
+
+    if (date.getDay() === 6) {
+      console.log(displayDate);
+    } else {
+      process.stdout.write(displayDate);
+      process.stdout.write(" ");
+    }
+  }
+
+  console.log();
```
この実装だと月末の日付が土曜日の場合に余分な空行が出力されそうです。

yokomaru replied:
上記、[こちらでコメントいただいた箇所](https://github.com/yokomaru/js-practices/pull/2#discussion_r1852276558)と合わせて以下のように修正いたしました。

・土曜日の場合は改行し`continue`してその後の処理をスキップ
・土曜以外の場合は日付を表示した後、月末であれば改行、そうでなければ空白を表示するように修正いたしました。

[8365217](https://github.com/yokomaru/js-practices/pull/2/commits/8365217ba9c6a696ddfbc8423b5c5ff566a88093)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,85 @@
+      console.log(displayDate);
+    } else {
+      process.stdout.write(displayDate);
+      process.stdout.write(" ");
```
この実装だと月末の土曜日でない日付の後ろにも余分な空白が追加されてしまいます。

yokomaru committed:
fix: update npm chalk version

- chalkのバージョンを4.1.2から最新の5.3.0に変更

yokomaru committed:
fix: change from bracket notation to dot notation

- argv["y"]をargv.yに変更
- argv["m"]をargv.mに変更

yokomaru committed:
fix: make if with block

- if文を使用する際はブロックで囲む

yokomaru committed:
fix: change variable name firstMonthDate

- 月の初日を表現するために、firstMonthDateからmonthFirstDateに変更

yokomaru committed:
fix: delete template string

- 不要なテンプレート文字列を削除

yokomaru committed:
add: monthFirstDay

- ループの都度monthFirstDate.getDayを取得していた箇所を変数に入れる

yokomaru committed:
fix: get the last day of the month

- 最終日までのループを次の月の1日を使用していたが、月の最終日でループするように修正
- 取得した最終日を変数にセット(monthLastDate)

yokomaru committed:
fix: split function init

- カレンダー表示処理の実行を切り出す
- バリデーションと値をセットする処理を分ける
- ネストしている条件を分けて、早期リターンするように修正
    - return console.errorしていた箇所をprocess.exit(1)でプロセスを止めるように修正する

yokomaru committed:
fix: rename buildCalender

- カレンダーを表示する処理のため、printCalenderに変更

yokomaru committed:
fix: printCalender

- 日付オブジェクトでループしていたが、最終日の日付の数字までループさせるように修正
    - ループの中で日付を取得し、今日の日付と比較を行い色をつける
- 土曜日であれば改行を行い以降の処理をスキップ
    - 日付が月末だったら改行し、そうでなければ空白を表示する

yokomaru committed:
fix: rename currentDate to today

- month、yearと合わせるために本日日付の変数をtodayに変更

yokomaru committed:
fix:  validate after settinng variables

- undefinedかどうかのチェックをバリデーション処理と変数セットの両方で行っていたため、先に変数をセットした上でバリデーション処理を実行して重複を削除する

yokomaru committed:
delete: calculate clender title spaces

- 計算で求めていた余白のスペースは、現状の仕様では必ず6にため計算を削除
- 不要になった変数も削除

yokomaru commented:
レビューありがとうございます。
いただいたレビューに加え、設計を見直しました。

## 元々の実装(処理の流れ)
- init()内に以下処理が入っている
    - オプション引数のパース
    - バリデーション
    - 値のセット
    - buildCalenderの呼び出し
- buildCalenderは受け取った引数(年と月と今日日付)でカレンダーの表示を行う
- init()を実行

## 修正後の実装(処理の流れ)
- 受け取った引数をパース
- setMonth、setYearで月と年の値をセット(引数があればその値、なければ今日日付からセット)
- validateMonth、validateYearでセットした変数の値のチェック
- buildCalenderにて受け取った引数(年と月と今日日付)でカレンダーの表示を行う

改めてご確認の程よろしくお願いいたします。

yokomaru committed:
fix: stores padded values in paddedDate

- 数字をテンプレート文字を使用して文字列に変換している処理を.toString()を使用して文字列に変換する
- 文字列に変換し文字詰する処理が重複していたため、結果を変数(paddedDate)に入れる

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,74 @@
+import minimist from "minimist";
+import chalk from "chalk";
+
+function setMonth(today, argvMonth) {
```
set とついている関数は、普通は何らかの値の設定を行う戻り値のない関数です。

yokomaru replied:
setは戻り値がなく、値自体を変化させるような副作用のある処理につけることを理解いたしました。
上記の処理に関しまして、[Null合体演算子](https://github.com/yokomaru/js-practices/pull/2/commits/1a75ac927da67e7d2ed0b49a186d100de0c2d51b)を使用することで処理がかなり短くなったため、見通しをよくするために関数自体を削除いたしました。

[1a75ac9](https://github.com/yokomaru/js-practices/pull/2/commits/1a75ac927da67e7d2ed0b49a186d100de0c2d51b)

yokomaru replied:
今回は処理自体を削除してしまったのですが、このような**変数に代入するための戻り値を返すような処理**場合、どのような命名が適切かご意見をお伺いしたいです。
変数に入れる値を取得している + 戻り値がある処理なので、`get`が適切かと思ったのですが、そもそも`get`は意味が曖昧になりあまり使うべきではないという意見も目にしたことがあり、いまいち自信が持てません。(`assign`や`create`も検討しました)

cafedomancer replied:
get でよいと思います。assign は set と同じなのでダメです。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,74 @@
+import chalk from "chalk";
+
+function setMonth(today, argvMonth) {
+  return typeof argvMonth === "undefined" ? today.getMonth() + 1 : argvMonth;
```
`undefined` のときに処理を行いたいならもっと簡潔な別の書き方できると思います。

yokomaru replied:
ありがとうございます。調べたところ[Null 合体演算子 \(??\)](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing)を使得そうだったので修正いたしました！
[1a75ac9](https://github.com/yokomaru/js-practices/pull/2/commits/1a75ac927da67e7d2ed0b49a186d100de0c2d51b)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,74 @@
+  return typeof argvYear === "undefined" ? today.getFullYear() : argvYear;
+}
+
+function validateMonth(argvMonth) {
```
validate は検証するという意味なので、検証した結果として boolean が返ってきそうな気がしますが、そうなっていないので混乱します。

yokomaru replied:
https://github.com/yokomaru/js-practices/pull/2#discussion_r1870429631 でいただいたコメント合わせて、validateに関する処理を修正いたしました。

- 変数に値をセットする前に引数の値を検証する
- validateMonth ValidateYearはbooleanを返すように修正し、関数の結果で`process.exit(1)`を行うようにする

[25b8e69](https://github.com/yokomaru/js-practices/pull/2/commits/25b8e69bcf6ebb45f0282beba2b4bf851db9c000)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,74 @@
+const month = setMonth(today, argv.m);
+const year = setYear(today, argv.y);
+validateMonth(month);
+validateYear(year);
```
set よりも先に validate が行われるべきじゃないですかね。`today` から得られた値を validate してもしょうがないはずです。validate すべきは `argv` から得られる値です。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,74 @@
+  console.log(`      ${month}月 ${year}`);
+  console.log("日 月 火 水 木 金 土");
+
+  const monthFirstDay = new Date(year, month - 1, 1).getDay();
```
オブジェクトのプロパティはオブジェクトから切り離さずにオブジェクトのまま扱ってください。https://bootcamp.fjord.jp/pages/utilize-objects

yokomaru replied:
こちらで対応いたしました。

[4e1616b](https://github.com/yokomaru/js-practices/pull/2/commits/4e1616b1ae5f098845b3cbd46ddb709b408eabb7)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,74 @@
+  const monthLastDate = new Date(year, month, 0).getDate();
+
+  for (let date = 1; date <= monthLastDate; date++) {
+    const calenderDate = new Date(year, month - 1, date);
```
Date オブジェクトをループの度に生成する必要があるのであれば、そもそもループカウンタを Date オブジェクトとするような実装にしてください。それから、現状のループカウンタの `date` には number が格納されているのに、この `calendarDate` には Date オブジェクトが格納されていて、異なるデータを格納する変数に同じような名前が付けられているため混乱します。

yokomaru replied:
日付型でループを行うように修正いたしました。
カウンタ変数 `date` を日付型にして、処理の後にカウンタ変数自体を+1日してループするようにしております。
また、異なるデータを格納する変数名の命名を混同させないように気をつけます。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,74 @@
+
+    if (calenderDate.getDay() === 6) {
+      console.log(displayDate);
+      continue;
```
`continue` を使う必要はないような。`continue` や `break` は逐次実行の流れを無視する命令で、プログラムの流れが分かりづらくなるので、本当に必要な場合 (残りの処理をスキップすることでパフォーマンスが改善するとか) 以外は使わない方がよいと思います。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,74 @@
+        ? chalk.bgWhite(paddedDate)
+        : paddedDate;
+
+    if (calenderDate.getDay() === 6) {
+      console.log(displayDate);
+      continue;
+    }
+
+    process.stdout.write(displayDate);
```
使う関数は違えど `displayDate` を標準出力に出力するという点では 59 行目と 63 行目の処理は重複しているように思います。読みやすさの観点からも `displayDate` の出力は一箇所に集約したほうがよいと思います。

yokomaru replied:
https://github.com/yokomaru/js-practices/pull/2#discussion_r1870440630 でご指摘いただいた内容と合わせて以下対応いたしました。

- `continue`を使用せず、条件分岐で改行 or 空白の表示を行う
- ` process.stdout.write`内で分岐を行い、処理を１つにまとめる

[bc8aceb](https://github.com/yokomaru/js-practices/pull/2/commits/bc8aceb2e53e1a594cb4d32528b2c8627fefa44b)

yokomaru replied:
上記、処理が横に長くなりすぎたので関数に切り出しました

[563e44a](https://github.com/yokomaru/js-practices/pull/2/commits/563e44a5d25c8aeb26ce87db761294a713284aa4)

yokomaru committed:
fix: user Nullish coalescing operator

- "undefined" と比較していた箇所をNull合体演算子を使用するように修正

yokomaru committed:
fix: use getDay() property of date object

- 曜日プロパティの数値を変数にいれて使っていたが以下を参考にオブジェクト型のまま使用するように修正

https://bootcamp.fjord.jp/pages/utilize-objects

yokomaru committed:
fix: delete setMonth and setYear

- 1a75ac9 の修正で処理が短くなったので、見通しを良くするために関数を削除する

yokomaru committed:
fix: return boolean in  validate method and stop processing at the result

- validateメソッドで成否を返却せず、メソッド内で処理の停止を行っている
    - メソッドで条件に合っているかをtrue/falseを返すように修正
    - 年月どちらかでもfalseの場合時処理を停止するように修正

yokomaru committed:
fix:  loop by date type

- 数値でループ→数字を元にnew Date()して使用していたので、日付でループするように修正
    -  for分のinitializeでnew dateを行い、カウンタ変数を都度+1日していく
- calenderDateを使用していた箇所をdateに変更

yokomaru committed:
fix: Combine date display in one place

- 条件により表示のメソッドを変更していたが、表示する処理は１つにまとめる
    - 土曜日または最終日の場合は改行、それ以外は空白を表示する
    - console.logは使わずにprocess.stdout.writeを使用

yokomaru committed:
fix: display blank for days

- forを使用して月初日の曜日の数値分空白を表示していた箇所をrepeat関数を使用して表示するように修正した

yokomaru committed:
fix: left and right sides of comparison today and date

- todayの値とdateが一致するかを比較していくため、左辺と右辺を入れ替える

yokomaru committed:
fix: make minor corrections regarding the year and month.

- validateYear内でargvMonthを使用していたため修正
- year変数に言える値をargv.mとしていたため修正

yokomaru committed:
fix: fix position to declare today

- today変数を宣言する箇所を使用する箇所の近くにまとめる

yokomaru committed:
fix: change the function declaration to an Arrow function

- モダンなJSの書き方に修正するため、functionで宣言している関数をアロー関数に修正

yokomaru committed:
add: make a function checking month last day or saturday

- 土曜か月末最終日かどうかをチェックする処理を関数に切り出す

yokomaru commented:
コメントありがとうございます！
指定いただいたコメントに加え、以下も追加で修正いたしました。

- functionで定義していた関数をアロー関数で書き直しました。
[9976a9f](https://github.com/yokomaru/js-practices/pull/2/commits/9976a9f95751cf1acb0295bc1499ac79e2e7a067)

- for分で開始の空白を表示していたところを、`repeat`関数の存在を知ったので書き直しました
 [c26106b](https://github.com/yokomaru/js-practices/pull/2/commits/c26106b38bfd45a38922a541280fe1555102398b)

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,83 @@
+    date <= monthLastDate;
+    date.setDate(date.getDate() + 1)
+  ) {
+    const paddedDate = date.getDate().toString().padStart(2, " ");
```
date の問題が解決していないです。https://github.com/yokomaru/js-practices/pull/2#discussion_r1870437990

yokomaru replied:
失礼いたしました。
**年月日**は`date`、**日にち**は`day`で使い分けるように修正いたします。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,83 @@
+  }
+};
+
+const isSaturdayOrMonthLastDate = (date, monthLastDate) => {
+  return (
+    date.getDay() === 6 || date.toDateString() === monthLastDate.toDateString()
+  );
+};
```
この関数を定義する意味がないような。関数に分かりやすい名前を付けることによって処理を抽象化でき、それによって呼び出し元のプログラムが分かりやすくなるとかなら意味があるのですが、`isSaturdayOrMonthLastDate` という名前は処理の内容をそのまま記述しただけでそうなっていないという点で意味がないです。ただし、関数をなくして処理をそのままバラせばよいというわけではないので注意してください。63 行目の読みづらさの問題がない状態を維持したまま修正してください。

yokomaru replied:
コメントありがとうございます。
上記、**関数は削除し、各判定の結果を変数に入れて分岐する** 対応をしようと思っております。
- 週末と月末の判定を無理に１つにまとめず別々に分ける
- 各々の判定ロジックはシンプルなので、関数には切り出さず結果を変数に入れて利用する

修正の方向性として認識合っていますでしょうか？
 
```js
const isEndOfWeek = date.getDay() === 6;
const isEndOfMonth = date.toDateString() === monthLastDate.toDateString();

process.stdout.write(
    `${displayDay}${isEndOfWeek || isEndOfMonth ? "\n" : " "}`,
);
```

yokomaru replied:
上記、提示していただいた以下の要件に一致した修正になっているかを確認したく質問いたしました。

> 関数をなくして処理をそのままバラせばよいというわけではない
> 63 行目の読みづらさの問題がない状態を維持したまま修正してください

今回の修正で関数は削除したものの、
説明変数に判定結果を代入したのでそのまま処理をバラした実装にはなっておらず、
関数化した際の読みやすさを保ったままにできているという認識なのですが、
なんとなく不安があり、方向性を確認したかったいう意図がありました🙇‍♀️

yokomaru committed:
fix: separate variable names by date and day

- 日付と日にちの変数名を分ける
    - dateは日付(Date型)
    - dayは日にち
- displayDayだと関数名のように見えるため、displayedDayに変更する

yokomaru committed:
fix: use Explanatory Variables

- 改行か空白かの分岐を読みやすくするために、以下を説明変数に代入する
    - 週末であるかどうかの判定
    - 月末日であるかどうかの判定
